Aaf 94 wishlist api read

psyduc opened the pull request.

psyduc committed:
add first route for families to read their wishlist

psyduc committed:
Add more error handling.

Add a family with an empty wishlist.

phuchle commented on server/routes/api.js:
```diff
@@ -104,4 +105,29 @@ routes.get('/pairing/paired', (req, res) => {
+routes.get('/wishlist/:familyId/read', (req, res) => {
+
+  const familyId = req.params.familyId;
+  if ( !familyId.match(/[0-F]{24}/gi) ) {
```
Can you add a comment describing what this regex does?

jwu910 replied:
~Does this regex catch lower cases too?~

rainrivas commented on server/routes/api.js:
```diff
@@ -7,6 +7,7 @@ const bodyParser = require('body-parser');
 const Donor = require('../models/donor');
 const Family = require('../models/family');
+const Wishlist = require('../models/wishlist');
 // Wishlist = Family.readWishlist();
```
Delete this comment since you've declared the variable above

phuchle commented on server/routes/api.js:
```diff
@@ -104,4 +105,29 @@ routes.get('/pairing/paired', (req, res) => {
+
+  const familyId = req.params.familyId;
+  if ( !familyId.match(/[0-F]{24}/gi) ) {
+    res.status(200).json({ message: 'No family found'});
```
If familyID doesn't pass the regex, I think we should return bad parameter instead of 'no family found'

rainrivas replied:
Shouldn't this be a 400 since it's a bad family id

rainrivas commented on server/routes/api.js:
```diff
@@ -104,4 +105,29 @@ routes.get('/pairing/paired', (req, res) => {
+        } else if (wishlist[0].list.length === 0) {
+          res.status(200).json({ message: 'No wishlist found.' });
+        } else {
+          res.status(200).json(wishlist[0].list);
```
Should we be adding a `{ message: "Wishlist found!" }` or just dumping the list here is good enough?

rainrivas reviewed (changes requested):
* Remove duplicated Wishlist declarartion
* Fix error handling on ln 112
*Question* Should we have the wishlist unit tests be created here or as a separate PR in sprint 3?
*Question* Are we going to to be doing BDD/TDD where we do the test scripts first and then code against that?

phuchle commented on server/routes/api.js:
```diff
@@ -104,4 +105,29 @@ routes.get('/pairing/paired', (req, res) => {
+        res.status(500).json(err);
+      } else {
+        if (wishlist.length === 0) {
+          res.status(200).json({ message: 'No family found.' });
```
This error message should be more descriptive.

phuchle commented on server/routes/api.js:
```diff
@@ -104,4 +105,29 @@ routes.get('/pairing/paired', (req, res) => {
+        if (wishlist.length === 0) {
+          res.status(200).json({ message: 'No family found.' });
+        } else if (wishlist[0].list.length === 0) {
+          res.status(200).json({ message: 'No wishlist found.' });
```
A more descriptive error message would help here as well.

phuchle reviewed (changes requested):
See comments.

psyduc committed:
Merge branch 'AAF-63-models' of github.com:5-gwoap/adopt-a-family into AAF-94-wishlist-api-read

jwu910 commented on server/routes/api.js:
```diff
@@ -104,4 +105,29 @@ routes.get('/pairing/paired', (req, res) => {
+    .exec((err, wishlist) => {
+      if(err) {
+        res.status(500).json(err);
+      } else {
```
If everything that isn't an error, we can just break the error if block here and have everything else not wrapped within this same if. I dont think anything else after the if block gets processed if we move into a return statement.

example: 

```javascript
if (err) {
  return res.status(500).json(err);
}

// rest of code
```

psyduc committed:
AAF94 Add a description for the regex
Fix the logic for the if statement
Change the response status to 400 and made the output messages more discriptive

rainrivas commented on server/routes/api.js:
```diff
@@ -115,4 +115,32 @@ routes.get('/pairing/paired', (req, res) => {
+          res.status(500).json(err);
+        } else {
+          if (wishlist.length === 0) {
+            res.status(400).json({ message: 'There\'s no family with that ID' });
```
This should not be a 400 error, the response was good, but returned 0.

rainrivas commented on server/routes/api.js:
```diff
@@ -115,4 +115,32 @@ routes.get('/pairing/paired', (req, res) => {
+          if (wishlist.length === 0) {
+            res.status(400).json({ message: 'There\'s no family with that ID' });
+          } else if (wishlist[0].list.length === 0) {
+            res.status(400).json({ message: 'Family exists but has no wishlist' });
```
This should not be a 400 error, the response was good, but returned 0.

rainrivas reviewed (changes requested):
Correct status codes

psyduc committed:
AAF 94 Change the HTTP code for no family found to 404.

jwu910 reviewed (changes requested):
I think run a quick prettier format and we're good.

jwu910 reviewed (approved):
LGTM. Gucci good

jwu910 commented:
Note: We can refactor with moving the logic into the wishlist controllers later.

psyduc committed:
AAF-94 Fix uppercase for const variable

phuchle commented on server/routes/api.js:
```diff
@@ -115,4 +115,37 @@ routes.get('/pairing/paired', (req, res) => {
   });
 });
 
+routes.get('/wishlist/:familyId/read', (req, res) => {
```
The route slug should read `'/wishlist/:familyId'` only.

rainrivas replied:
good catch.. totally slipped me up

jwu910 commented on server/routes/api.js:
```diff
@@ -117,13 +117,13 @@ routes.get('/pairing/paired', (req, res) => {
     res.status(400).json({ message: 'Invalid family id'});
   } else {
-    Wishlist.find({family: familyId})
+    Wishlist.find({family: FAMILY_ID})
```
Can we discuss the use of CONSTANTS and assigning variables with const? 

I'm not seeing the need to capitalize these.

rainrivas replied:
We decided to make constants caps a long *long* time ago... I remember having this discussion when we were talking code conventions and deciding what rules the linter would catch.

jwu910 replied:
Yeah I have an understanding that constants should be all caps. but I dont see this being a semantic CONSTANT. 

To my understanding I would use all caps constants for something like a reusable string that we pre-define going in to the app. like "ORGANIZATION_NAME" and we'd define it as "Adopt-A-Family" and then we can call the constant ORGANIZATION_NAME anywhere in our code and it would reference that.

But we are binding a variable with const so it cannot be re-bound again during this function.

rainrivas replied:
So are you saying that this shouldn't be a constant and you think it should be some other type of variable?
I have only used constants in my tests for defining end points so that would be an example of a 'non-global' reusable string

jwu910 replied:
I am saying the use of non-reassignable variables in javascript do not need to be all caps. The only time I've actually seen it used as ALL_CAPS is when the value assigned to it is literally constant. Like my example earlier, a website name, a string for an action type.

If the value assigned to that variable comes from a user, from the return of a function, or a relative value, and we just don't want that being reassigned, I have only seen camelCased. 

In your tests:

```
const PAIRING_ENDPOINT = '/api/pairing';
const WISHLIST_ENDPOINT = '/api/wishlist';
const WISHLIST_CREATE_ENDPOINT = '/api/wishlist/addItem';
```

I agree this works because you're saying that these endpoints are going to be these and will never change. They're pre-defined and not a result of the frontend passing information in; or the result of a function call.

phuchle replied:
In this particular case, I think the line in question should be `Wishlist.find({family: familyId})` and all caps case is misleading as the value is not static.

rainrivas replied:
I hear you and I agree 50%. If we are only wanting this variable to be scoped here, why don't we use `let` since it's just a way to make `req.params.familyId` easier to read. In that case wouldn't a `let family = req.params.familyId` be the right way to go?
Would anyone else here care to chime in on this? Should we go with `const` or `let` and if so, should we use the `const` with camelCase since it's not a true constant as josh pointed out

jwu910 replied:
>is not static

Is probably a much more concise of saying what im saying.

phuchle replied:
It's fine to be a const declaration with camel case.  You could even destructure like so: 
`const { familyId } = req.params;`

jwu910 replied:
destructuring is super cool

rainrivas replied:
^I like that a lot. Makes it very clear that we're pulling a "constant" from the request.

rainrivas reviewed (changes requested):
Correct route requested by Phuc

psyduc committed:
AAF-94 Revert the change to the familyId variable and fix the route

rainrivas commented:
I notice that the delete code somehow made it into this branch... that's kinda bad considering that code isn't clean and ready to be in here, but whatever; we can fix it with the next pr

psyduc committed:
AAF-94 Remove the delete route

psyduc commented:
Whoops! The delete route was totally an accident. Let me remove before we merge. @rainrivas
